src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_libuv.cc

```diff
@@ -0,0 +1,132 @@
+#include <grpc/support/string_util.h>
+#include <grpc/support/time.h>
+#include "src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h"
+#include "src/core/lib/gpr/string.h"
+
+namespace grpc_core {
+
+void ares_uv_poll_cb(uv_poll_t* handle, int status, int events);
+
+void ares_uv_poll_close_cb(uv_handle_t* handle) { Delete(handle); }
+
+class GrpcPolledFdLibuv : public GrpcPolledFd {
+ public:
+  GrpcPolledFdLibuv(ares_socket_t as) : as_(as) {
+    gpr_asprintf(&name_, "c-ares socket: %" PRIdPTR, (intptr_t)as);
+    handle_ = New<uv_poll_t>();
+    uv_poll_init_socket(uv_default_loop(), handle_, as);
+    handle_->data = this;
+  }
+
+  ~GrpcPolledFdLibuv() { gpr_free(name_); }
+
+  void RegisterForOnReadableLocked(grpc_closure* read_closure) override {
+    GPR_ASSERT(read_closure_ == nullptr);
+    GPR_ASSERT((poll_events_ & UV_READABLE) == 0);
+    read_closure_ = read_closure;
+    poll_events_ |= UV_READABLE;
+    uv_poll_start(handle_, poll_events_, ares_uv_poll_cb);
+  }
+
+  void RegisterForOnWriteableLocked(grpc_closure* write_closure) override {
+    GPR_ASSERT(write_closure_ == nullptr);
+    GPR_ASSERT((poll_events_ & UV_WRITABLE) == 0);
+    write_closure_ = write_closure;
+    poll_events_ |= UV_WRITABLE;
+    uv_poll_start(handle_, poll_events_, ares_uv_poll_cb);
+  }
+
+  bool IsFdStillReadableLocked() override {
+    /* uv_poll_t is based on poll, which is level triggered. So, if cares
+     * leaves some data unread, the event will trigger again. */
+    return false;
+  }
+
+  void ShutdownLocked(grpc_error* error) override {
+    uv_poll_stop(handle_);
+    uv_close(reinterpret_cast<uv_handle_t*>(handle_), ares_uv_poll_close_cb);
+  }
+
+  ares_socket_t GetWrappedAresSocketLocked() override { return as_; }
+
+  const char* GetName() override { return name_; }
+
+  char* name_;
+  ares_socket_t as_;
+  uv_poll_t* handle_;
+  grpc_closure* read_closure_ = nullptr;
+  grpc_closure* write_closure_ = nullptr;
+  int poll_events_ = 0;
+};
+
+void ares_uv_poll_cb(uv_poll_t* handle, int status, int events) {
+  grpc_core::ExecCtx exec_ctx;
+  GrpcPolledFdLibuv* polled_fd =
+      reinterpret_cast<GrpcPolledFdLibuv*>(handle->data);
+  grpc_error* error = GRPC_ERROR_NONE;
+  if (status < 0) {
+    error = GRPC_ERROR_CREATE_FROM_STATIC_STRING("cares polling error");
+    error =
+        grpc_error_set_str(error, GRPC_ERROR_STR_OS_ERROR,
+                           grpc_slice_from_static_string(uv_strerror(status)));
+  }
+  if ((events & UV_READABLE) && polled_fd->read_closure_) {
+    GRPC_CLOSURE_SCHED(polled_fd->read_closure_, error);
+    polled_fd->read_closure_ = nullptr;
+    polled_fd->poll_events_ &= ~UV_READABLE;
+  }
+  if ((events & UV_WRITABLE) && polled_fd->write_closure_) {
+    GRPC_CLOSURE_SCHED(polled_fd->write_closure_, error);
+    polled_fd->write_closure_ = nullptr;
+    polled_fd->poll_events_ &= ~UV_WRITABLE;
+  }
+  uv_poll_start(handle, polled_fd->poll_events_, ares_uv_poll_cb);
```

Comment:
At first glance, I'm wondering if this is a use-after-close on `handle`. Is there some guarantee that the `uv_close` callback will run only after any outstanding `uv_poll_start` callbacks have ran? Further, in that case is there a guarantee that `uv_poll_start` is a no-op is there's a pending `uv_close`?

Reply:
The guarantee is that the callback will not be run after `uv_poll_stop` is called. And we call that before calling `uv_close`. So, this line cannot be reached after the handle is closed.

Reply:
Note that after grpc calls `RegisterForOnReadable/Writeable`, the `GrpcPolledFd` refs the object driving overall resolution, and that ref won't release until the passed-in closure is called. Therefore, the `GrpcPolledFd` `ShutdownLocked` method needs to be able to cancel any outstanding callbacks. Otherwise cancellation of DNS resolution, e.g. on a timeout, won't work (there are other cases that rely on this too e.g. fallback from UDP to TCP).

If this callback can't be ran after `uv_poll_stop` is called, then it sounds like this won't work, i.e. there's not a way to cancel pending reads and writes?

Reply:
OK, I didn't follow at first. Would it work to just cancel those closures in the `ShutdownLocked` method? Also, what exactly does canceling a closure entail? Do I have to call it with a specific error?

Reply:
The closure should just be scheduled with some grpc error that is not `GRPC_ERROR_NONE`.

Overall yes, I think that would work. The only remaining concern I'd have is that the work to fully close the handlers is done, but it sounds like the libuv poller in the background is taking care of that.

Reply:
I have added this.